traffic_ops/traffic_ops_golang/topology/topologies.go

```diff
@@ -0,0 +1,485 @@
+	}
+	if err = rows.Close(); err != nil {
+		userErr, sysErr, errCode := api.ParseDBError(err)
+		return nil, userErr, sysErr, errCode
+	}
+
+	for _, topology := range topologies {
+		nodeMap := map[int]int{}
+		for index, node := range topology.Nodes {
+			nodeMap[node.Id] = index
+		}
+		for _, node := range topology.Nodes {
+			for parentIndex := 0; parentIndex < len(node.Parents); parentIndex++ {
+				node.Parents[parentIndex] = nodeMap[node.Parents[parentIndex]]
+			}
+		}
+		interfaces = append(interfaces, *topology)
+	}
+	return interfaces, nil, nil, http.StatusOK
+}
+
+func (topology *TOTopology) removeParents() error {
+	_, err := topology.ReqInfo.Tx.Exec(deleteParentsQuery(), topology.Name)
+	if err != nil {
+		return errors.New("topology update: error deleting old parents: " + err.Error())
+	}
+	return nil
+}
+
+func (topology *TOTopology) removeNodes(cachegroups *[]string) error {
+	_, err := topology.ReqInfo.Tx.Exec(deleteNodesQuery(), topology.Name, pq.Array(*cachegroups))
+	if err != nil {
+		return errors.New("topology update: error removing old unused nodes: " + err.Error())
+	}
+	return nil
+}
+
+func (topology *TOTopology) addNodes() (error, error, int) {
+	var cachegroupsToInsert []string
+	var indices = make([]int, 0)
+	for index, node := range topology.Nodes {
+		if node.Id == 0 {
+			cachegroupsToInsert = append(cachegroupsToInsert, node.Cachegroup)
+			indices = append(indices, index)
+		}
+	}
+	if len(cachegroupsToInsert) == 0 {
+		return nil, nil, http.StatusOK
+	}
+	rows, err := topology.ReqInfo.Tx.Query(nodeInsertQuery(), topology.Name, pq.Array(cachegroupsToInsert))
+	if err != nil {
+		return nil, errors.New("error adding nodes: " + err.Error()), http.StatusInternalServerError
+	}
+	for _, index := range indices {
+		rows.Next()
+		err = rows.Scan(&topology.Nodes[index].Id, &topology.Name, &topology.Nodes[index].Cachegroup)
+		if err != nil {
+			return api.ParseDBError(err)
+		}
+	}
+	if err = rows.Close(); err != nil {
+		return api.ParseDBError(err)
+	}
+	return nil, nil, http.StatusOK
+}
+
+func (topology *TOTopology) addParents() (error, error, int) {
+	var (
+		children []int
+		parents  []int
+		ranks    []int
+	)
+	for _, node := range topology.Nodes {
+		for rank := 1; rank <= len(node.Parents); rank++ {
+			parent := topology.Nodes[node.Parents[rank-1]]
+			children = append(children, node.Id)
+			parents = append(parents, parent.Id)
+			ranks = append(ranks, rank)
+		}
+	}
+	rows, err := topology.ReqInfo.Tx.Query(nodeParentInsertQuery(), pq.Array(children), pq.Array(parents), pq.Array(ranks))
+	if err != nil {
+		return api.ParseDBError(err)
+	}
+	for _, node := range topology.Nodes {
+		for rank := 1; rank <= len(node.Parents); rank++ {
+			rows.Next()
+			parent := topology.Nodes[node.Parents[rank-1]]
+			err = rows.Scan(&node.Id, &parent.Id, &rank)
+			if err != nil {
+				return api.ParseDBError(err)
+			}
+		}
+	}
+	err = rows.Close()
```

Comment:
closing the rows needs to be in a `defer` in order to ensure that the rows are properly closed however this method may exit

Reply:
Deferring in 73fb323465.